### PR TITLE
Migrate web3

### PR DESCRIPTION
**Pull request by jefflau**

### CODE REVIEW

File: src/api/rootResolver.js [hunk 1]
```diff
@@ -27,8 +30,8 @@ const resolvers = {
     async accounts() {},
     async ethers() {
```
Comment (hiddentao) [inline src/api/rootResolver.js]: ethers?

File: src/api/web3.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+import Web3 from 'web3'
+
+import { DEPLOYER_CONTRACT_ADDRESS, NETWORK } from '../config'
+
+let web3
+export let networkError
+
+function getWeb3() {
+  if (!web3) {
+    return setupWeb3()
```
Comment (hiddentao) [inline src/api/web3.js]: `setupWeb3()` returns a promise. I think that `getWeb3()` should just return `web3`. Clients calling should take appropriate action if it's not set.

I would rewrite this as:

```js
if (!web3) {
  setupWeb3()
}
return web3
```

File: src/api/web3.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+export async function getDeployerAddress() {
+  // if local env doesn't specify address then assume we're on a public net
+  const id = await web3.eth.net.getId()
+  return DEPLOYER_CONTRACT_ADDRESS || Deployer.NETWORKS[id].address
+}
+
+export async function getTransactionLogs(txHash) {
+  const web3 = getWeb3()
+
+  const logs = await web3.eth.getTransactionReceipt(txHash)
```
Comment (hiddentao) [inline src/api/web3.js]: can do: `return web3.eth.get....`

File: src/api/web3.js [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+export async function getAccount() {
+  const web3 = getWeb3()
+  const accounts = await web3.eth.getAccounts()
+  return accounts[0]
+}
+
+export async function setupWeb3() {
+  //Localnode
```
Comment (hiddentao) [inline src/api/web3.js]: I don't think we should try to connect directly to `localhost`. That's what a dapp browser / metamask is for. Just switch networks within Metamask - this way we also ensure we're testing our Metamask link up locally.

I think the flow we want is:

1. Check if web3.currentProvider is available. If so then use that.
2. If not then use the Infura read-only endpoint.

File: src/api/web3.js [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+  if (web3) {
+    return web3
+  }
+
+  // Checking if Web3 has been injected by the browser (Mist/MetaMask)
+  if (window.web3 && window.web3.currentProvider) {
+    web3 = new Web3(window.web3.currentProvider)
+    console.log(web3)
+    return web3
```
Comment (hiddentao) [inline src/api/web3.js]: Dont' need to return it since `getWeb3()` above will be the accessor

File: src/api/web3.js [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+      }
+    })
+
+  if (web3) {
+    return web3
+  }
+
+  // Checking if Web3 has been injected by the browser (Mist/MetaMask)
+  if (window.web3 && window.web3.currentProvider) {
+    web3 = new Web3(window.web3.currentProvider)
```
Comment (hiddentao) [inline src/api/web3.js]: Instantiating is insufficient - we should do a `getNetwork()` call in here to ensure the endpoint actually works.

File: src/api/web3.js (hunk 4 above)
Comment (hiddentao) [reply src/api/web3.js]: Also, the config JSON will specify which network is expected. The setup method to check that we're connected to the expected network.

File: src/routes/SingleEventAdmin.js [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    this.setState({
+      search: event.target.value
+    })
+  }
+
+  render() {
+    const { address } = this.props.match.params
+    return (
+      <SingleEventContainer>
+        <Query query={PartyQuery} variables={{ address }}>
```
[the remark this answers is not in the document]
Comment (jefflau) [reply src/routes/SingleEventAdmin.js]: Will think about how we do this later :) Just had this here cos it was outstanding from when I wanted to migrate

File: src/layout/Layouts.js [hunk 8]
```diff
@@ -3,7 +3,7 @@ import styled from 'react-emotion'
 import Header from './Header'
 import Footer from './Footer'
 
-import { networkError } from '../api/ethers'
+import { networkError } from '../api/web3'
```
Comment (hiddentao) [inline src/layout/Layouts.js]: Need to use a function otherwise this won't work

File: src/api/web3.js [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    try {
+      await web3.eth.net.getId()
+    } catch (e) {
+      networkError = `We were unable to connect to the Ethereum network`
+      console.error(networkError)
+    }
+  } else {
+    console.log('No web3 instance injected.')
+  }
+}
```
Comment (hiddentao) [inline src/api/web3.js]: What about fallback to infura?

File: src/routes/SingleEventAdmin.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (hiddentao) [reply src/routes/SingleEventAdmin.js]: Ok

File: src/api/web3.js [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+  if (window.web3 && window.web3.currentProvider) {
+    web3 = new Web3(window.web3.currentProvider)
+    try {
+      await web3.eth.net.getId()
+    } catch (e) {
+      networkError = `We were unable to connect to the Ethereum network`
+      console.error(networkError)
+    }
+  } else {
+    web3 = new Web3('https://mainnet.infura.io/')
```
Comment (hiddentao) [inline src/api/web3.js]: The inufra endpoint has to be based on the `config.NETWORK` setting, so if no NETWORK is specified then obviously at this point we error out and we can't connect.

Also, infura endpoint should be tested too (net.getId()).

File: src/api/web3.js [hunk 11]
```diff
@@ -0,0 +1,69 @@
+import { Deployer } from '@noblocknoparty/contracts'
+import Web3 from 'web3'
+
+import { DEPLOYER_CONTRACT_ADDRESS, NETWORK } from '../config'
+
+let web3
+export let networkError
```
Comment (hiddentao) [inline src/api/web3.js]: no need to export anymore

**Pull request merged**